astropy/utils/column_info.py

```diff
@@ -0,0 +1,232 @@
+import os
+import sys
+from copy import deepcopy
+import weakref
+import numpy as np
+from functools import partial
+
+from ..extern import six
+from ..utils import OrderedDict
+
+COLUMN_ATTRS = set(['name', 'unit', 'dtype', 'format', 'description', 'meta', 'parent_table'])
+INFO_SUMMARY_ATTRS = ('dtype', 'shape', 'unit', 'format', 'description', 'class')
+
+
+def column_info_factory(names, funcs):
+    """
+    Factory to create a function that can be used as an ``option``
+    for outputting table or column summary information.
+
+    Examples
+    --------
+    >>> from astropy.table.info import column_info_factory
+    >>> from astropy.table import Column
+    >>> c = Column([4., 3., 2., 1.])
+    >>> mystats = column_info_factory(names=['min', 'median', 'max'],
+    ...                               funcs=[np.min, np.median, np.max])
+    >>> c.info(option=mystats)
+    min = 1.0
+    median = 2.5
+    max = 4.0
+
+    Parameters
+    ----------
+    names: list
+        List of information attribute names
+    funcs: list
+        List of functions that compute the corresponding information attribute
+
+    Returns
+    -------
+    func: function
+        Function that can be used as a table or column info option
+    """
+    def func(col):
+        outs = []
+        for name, func in zip(names, funcs):
+            try:
+                if isinstance(func, six.string_types):
+                    out = getattr(col, func)()
+                else:
+                    out = func(col)
+            except:
+                outs.append('--')
+            else:
+                outs.append(str(out))
+
+        return OrderedDict(zip(names, outs))
+    return func
+
+
+def _get_column_attribute(col, attr=None):
+    """
+    Get a column attribute for the ``attributes`` info summary method
+    """
+    from ..table.column import BaseColumn
+
+    if attr == 'class':
+        val = '' if isinstance(col, BaseColumn) else col.__class__.__name__
+    elif attr == 'dtype':
+        val = col.info.dtype.name
+    elif attr == 'shape':
+        colshape = col.shape[1:]
+        val = colshape if colshape else ''
+    else:
+        val = getattr(col.info, attr)
+    if val is None:
+        val = ''
+    return str(val)
+
+
+class ColumnInfo(object):
```

Comment:
Actually, maybe even simpler, if we do go for `DataInfo` here, is to just remove the check on `class` altogether and in `column.py` construct a proper `ColumnInfo(DataInfo)` that special-cases `class`.